crates/ra_ide/src/display/structure.rs

```diff
@@ -151,10 +151,24 @@ fn structure_node(node: &SyntaxNode) -> Option<StructureNode> {
                 Some(node)
             },
             ast::MacroCall(it) => {
-                let first_token = it.syntax().first_token().unwrap();
-                if first_token.text().as_str() != "macro_rules" {
-                    return None;
+                let macro_name = it.syntax()
+                    .children()
```

Comment:
IIRC, `ast:MacroCall` implmenet `NameOwner`, we could use it to get the name directly.

Reply:
Any idea, how I can get a ast::MacroCAll?

Reply:
I use the following construct, which is more than ugly: `let n = ast::MacroCall {syntax: *it.syntax()}.name();`

Reply:
`it` itself should be `MacroCall`? right?

Reply:
No, does not seem like it works?

Reply:
Never mind, the error squiggles were not updated.

Reply:
NameOwner::name() returns the name of the macro, but not the macro invocation macro_rules!.

Maybe this is already handled somewhere else? Otherwise nameowner would not work on the macro definition. Is it possible to query HIR instead of the plain syntax tree in file_structure()?

Reply:
Seems like MacroCall uses the default impl for NameOwner::name(), which will just pick up the first thing you can cast to a Name.

Reply:
You need to look at the `.path` and `.name` of an `ast::MacroCall`: 

https://github.com/rust-analyzer/rust-analyzer/blob/19eb7fa1db7da8417314ddfafe7addbbd9c3b46a/crates/ra_hir_def/src/nameres/raw.rs#L375-L380

Reply:
It's path because `std::prelude::println!("hello")`

Reply:
In theory I should use Path::from_ast() right? It is deprecated because it does not handle $crate, but that is not relevant. Defining a macro inside a macro expansion is not possible as far as i know.